ground/src/main/java/com/google/android/ground/repository/SurveyRepository.kt

```diff
@@ -79,29 +82,32 @@ constructor(
 
   init {
     // Kicks off the loading process whenever a new survey id is selected.
+    // TODO(#1426): Simplify this flow by moving into function invocation.
     selectSurveyEvent
       .distinctUntilChanged()
-      .switchMap { selectSurvey(it) }
+      .switchMap { surveyId ->
+        if (surveyId.isEmpty()) Flowable.never()
+        else
+          rxSingle { loadSurvey(surveyId) }
+            .toFlowable()
+            .compose { survey -> Loadable.loadingOnceAndWrap(survey) }
+      }
       .onBackpressureLatest()
       .subscribe(surveyLoadingState)
   }
 
-  private fun selectSurvey(surveyId: String): @Cold Flowable<Loadable<Survey>> {
-    // Empty id indicates intent to deactivate the current survey or first login.
-    return if (surveyId.isEmpty()) Flowable.never()
-    else
-      remoteDataStore
-        .subscribeToSurveyUpdates(surveyId)
-        .andThen(
-          Flowable.defer {
-            syncSurveyWithRemote(surveyId)
-              .onErrorResumeNext { getSurvey(surveyId) }
-              .map { attachJobPermissions(it) }
-              .doOnSuccess { lastActiveSurveyId = surveyId }
-              .toFlowable()
-              .compose { Loadable.loadingOnceAndWrap(it) }
-          }
-        )
+  private suspend fun loadSurvey(surveyId: String): Survey {
+    var survey = surveyStore.getSurveyById(surveyId).awaitSingleOrNull()
+    if (survey == null) {
+      survey = syncSurveyFromRemote(surveyId)
+    }
+    this.lastActiveSurveyId = surveyId
+    return attachJobPermissions(survey)
+  }
+
+  private suspend fun syncSurveyFromRemote(surveyId: String): Survey {
+    remoteDataStore.subscribeToSurveyUpdates(surveyId).await()
```

Comment:
Why aren't we subscribing to updates when the survey exists locally?

Reply:
Check out the Javadoc for `subscribeToSurveyUpdates` (with some corrections):

```
  /**
   * Listens for remote changes to the survey with the specified id. Implementations should handle
   * synchronization of new, changed, and deleted surveys and LOIs to the local db in the
   * background.
   */
  fun subscribeToSurveyUpdates(surveyId: String): @Cold Completable
```

The impl uses Firebase Cloud Messaging to listen for pings to reload the survey and LOIs when they're updated. Implemented in #1408.